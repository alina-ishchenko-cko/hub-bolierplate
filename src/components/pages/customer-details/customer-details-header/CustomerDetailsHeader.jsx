// @flow
import * as React from 'react';
import { FlexColumn, FlexItem } from 'components/ui/flex';
import isEqual from 'lodash/isEqual';
import {
  WrapperStyled,
  CustomerNameWrapperStyled,
  CustomerInfoWrapperStyled,
} from './CustomerDetailsHeader.sc';
import CkoButton from 'components/ui/button';
import CkoIcon from 'components/ui/icon';
import UpdateCustomerDetails from '../update-customer-details';
import CopyToClipboardBtn from 'components/ui/copy-to-clipboard-btn';
import moment from 'moment';
import CkoLoading from 'components/ui/loading/';

type Props = {
  details: {
    name: string,
    createdDate: string,
    phone: Object,
    email: string,
    id: string,
  },
  isReadOnly: boolean,
  countries: Array<Object>,
  updateCustomerDetails: Function,
};

type State = {
  loading: boolean,
  isEditDetailsDialogVisible: boolean,
};

export default class CustomerDetailsHeader extends React.Component<
  Props,
  State
> {
  state = {
    loading: false,
    isEditDetailsDialogVisible: false,
  };

  shouldComponentUpdate(nextProps: Props, nextState: State) {
    return this.hasDataChanged(nextProps) || !isEqual(nextState, this.state);
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (this.hasDataChanged(prevProps)) {
      this.setState({ loading: true });
    } else if (this.state.loading) {
      // Add a bit of delay to make the loader icon more obvious
      setTimeout(() => {
        this.setState({ loading: false });
      }, 300);
    }
  }

  hasDataChanged(nextProps: Props): boolean {
    const prevDetails = {
      name: this.props.details.name,
      createdDate: this.props.details.createdDate,
      phone: this.props.details.phone,
      email: this.props.details.email,
      id: this.props.details.id,
    };
    const details = {
      name: nextProps.details.name,
      createdDate: nextProps.details.createdDate,
      phone: nextProps.details.phone,
      email: nextProps.details.email,
      id: nextProps.details.id,
    };
    return !isEqual(prevDetails, details);
  }

  toggleEditDetailsDialog = () => {
    this.setState(prevState => {
      return {
        isEditDetailsDialogVisible: !prevState.isEditDetailsDialogVisible,
      };
    });
  };

  getFormattedPhone() {
    // TODO find a library to format phone according to the country standards
    const { phone } = this.props.details;
    if (!phone || !phone.number) {
      return 'N/A';
    }

    let formattedPhone = '';
    if (phone.countryCode) {
      formattedPhone = `+${phone.countryCode} `;
    }
    if (phone.number) {
      formattedPhone = `${formattedPhone}${phone.number}`;
    }
    return formattedPhone;
  }

  render() {
    return (
      <WrapperStyled className="customer-details-header">
        {this.state.loading && <CkoLoading full opacity="1" />}
        {this.state.isEditDetailsDialogVisible ? (
          <UpdateCustomerDetails
            onCancel={this.toggleEditDetailsDialog}
            updateCustomerDetails={this.props.updateCustomerDetails}
            name={this.props.details.name}
            email={this.props.details.email}
            phone={this.props.details.phone}
            customerId={this.props.details.id}
            countries={this.props.countries}
          />
        ) : null}
        <FlexColumn>
          <CustomerNameWrapperStyled>
            <FlexItem className="customer-name">
              {this.props.details.name || 'N/A'}
            </FlexItem>
            {!this.props.isReadOnly ? (
              <FlexItem className="actions">
                <CkoButton
                  className="edit-details-btn"
                  size="large"
                  icon="edit"
                  value="Edit"
                  dataIndex="editCustomerDetails"
                  onClick={this.toggleEditDetailsDialog}
                />
              </FlexItem>
            ) : null}
          </CustomerNameWrapperStyled>
          <CustomerInfoWrapperStyled>
            <FlexColumn
              className="customer-info-column"
              alignItems="start"
              grow={0}>
              <FlexItem className="email">
                <CkoIcon className="email-icon" name="email" />
                <span> {this.props.details.email}</span>
              </FlexItem>
              <FlexItem className="create-date">
                <CkoIcon className="date-icon" name="date" />
                <span>
                  {moment(this.props.details.createdDate).format(
                    'DD/MM/YYYY HH:mm:ss'
                  )}
                </span>
              </FlexItem>
            </FlexColumn>
            <FlexColumn
              className="customer-info-column"
              alignItems="start"
              grow={0}>
              <FlexItem className="phone">
                <CkoIcon className="phone-icon" name="phone" />
                <span>{this.getFormattedPhone()}</span>
              </FlexItem>
              <FlexItem className="plan-id">
                <CkoIcon className="id-icon" name="id" />
                <span className="id-text">{this.props.details.id}</span>
                <CopyToClipboardBtn text={this.props.details.id} />
              </FlexItem>
            </FlexColumn>
          </CustomerInfoWrapperStyled>
        </FlexColumn>
      </WrapperStyled>
    );
  }
}
