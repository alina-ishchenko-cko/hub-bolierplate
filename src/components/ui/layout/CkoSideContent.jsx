// @flow
import * as React from 'react';
import CkoIcon from 'components/ui/icon/';
import CkoLoading from 'components/ui/loading/';
import isEqual from 'lodash/isEqual';
import { Scrollbars } from 'react-custom-scrollbars';
import { FlexItem, FlexColumn } from 'components/ui/flex/';
import {
  ContentStyled,
  ContainerStyled,
  CloseBtnStyled,
  HeaderStyled,
} from './styled/CkoSideContent.sc';

type Props = {
  loading?: boolean,
  className?: string,
  header: React.Node,
  onClose: Function,
  children: React.Node,
  showBackBtn: boolean,
};

type State = {
  uiLoading: boolean,
};

export default class CkoSideContent extends React.Component<Props, State> {
  timer: any;
  state = {
    uiLoading: false,
  };

  componentWillReceiveProps(nextProps: Props) {
    if (!isEqual(this.props.header, nextProps.header)) {
      this.showUiLoading();
    }
  }

  showUiLoading() {
    // Clear the previous time
    clearTimeout(this.timer);

    // Show Loader
    this.setUiLoading(true);

    // Scroll the content area to the top
    //this.refs.scrollWrap.scrollTop(0);

    // Add 300ms delay to show loader
    this.timer = setTimeout(this.setUiLoading, 300);
  }

  setUiLoading = (uiLoading?: boolean = false) => {
    this.setState({ uiLoading });
  };

  handleClose = (e: SyntheticEvent<any>) => {
    e.preventDefault();
    e.stopPropagation();
    this.props.onClose();
  };

  isLoading() {
    return this.props.loading || this.state.uiLoading;
  }

  render() {
    const iconType = this.props.showBackBtn ? 'right' : 'close';
    return (
      <ContainerStyled
        className={`cko-side-content ${this.props.className || ''}`}>
        <CloseBtnStyled onClick={this.handleClose}>
          <CkoIcon name={iconType} />
        </CloseBtnStyled>
        {this.props.loading && <CkoLoading full opacity="1" />}
        <FlexColumn className="cko-content-wrap">
          <FlexItem grow="0">
            <HeaderStyled>{this.props.header}</HeaderStyled>
          </FlexItem>
          <Scrollbars>
            <ContentStyled>{this.props.children}</ContentStyled>
          </Scrollbars>
        </FlexColumn>
      </ContainerStyled>
    );
  }
}
