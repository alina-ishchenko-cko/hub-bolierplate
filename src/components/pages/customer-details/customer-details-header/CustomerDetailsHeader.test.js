import React from 'react';
import CustomerDetailsHeader from './CustomerDetailsHeader';
import CopyToClipboardBtn from 'components/ui/copy-to-clipboard-btn';
import UpdateCustomerDetails from '../update-customer-details/';
import { HeaderStyled } from 'components/ui/layout/styled/CkoSideContent.sc';
import {
  CustomerNameWrapperStyled,
  CustomerInfoWrapperStyled,
} from './CustomerDetailsHeader.sc';

describe('Customer Details Header', () => {
  let wrapper;
  let mockProps;

  beforeEach(() => {
    mockProps = {
      details: {
        name: 'Alice',
        createdDate: '2018-04-17T13:51:49.837',
        phone: {},
        email: 'a@a.com',
        id: 'plan_1',
      },
      isReadOnly: false,
      countries: [
        {
          countryId: 1,
          countryIso2Code: 'AF',
          countryPhoneCode: '93',
          name: 'Afghanistan',
        },
        {
          countryId: 239,
          countryIso2Code: 'AX',
          countryPhoneCode: '358 18',
          name: 'Aland',
        },
      ],
      updateCustomerDetails: jest.fn(),
    };
    wrapper = shallow(<CustomerDetailsHeader {...mockProps} />);
    wrapper.setState({ loading: false, isEditDetailsDialogVisible: false });
  });

  // beforeEach(() => {
  //   wrapper.setProps(mockProps);
  //   wrapper.setState({ loading: false, isEditDetailsDialogVisible: false });
  // });

  it('renders without crashing', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<CustomerDetailsHeader {...mockProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains all required components', () => {
    expect(wrapper.find(CustomerNameWrapperStyled)).toHaveLength(1);
    expect(wrapper.find(CustomerInfoWrapperStyled)).toHaveLength(1);
    expect(wrapper.find('.customer-name')).toHaveLength(1);
    expect(wrapper.find('.edit-details-btn')).toHaveLength(1);
    expect(wrapper.find('.email')).toHaveLength(1);
    expect(wrapper.find('.create-date')).toHaveLength(1);
    expect(wrapper.find('.phone')).toHaveLength(1);
    expect(wrapper.find('.plan-id')).toHaveLength(1);
    expect(wrapper.find(CopyToClipboardBtn)).toHaveLength(1);
  });

  it('does not show Edit button if user is read only', () => {
    wrapper.setProps({
      isReadOnly: true,
    });
    expect(wrapper.find('button.edit-details-btn')).toHaveLength(0);
  });

  it('toggles update details dialog if isEditDetailsDialogVisible changes', () => {
    wrapper.setState({ isEditDetailsDialogVisible: true });
    expect(wrapper.find(UpdateCustomerDetails)).toHaveLength(1);
    wrapper.setState({ isEditDetailsDialogVisible: false });
    expect(wrapper.find(UpdateCustomerDetails)).toHaveLength(0);
  });

  it('toggles state on Edit button click', () => {
    const prevIsEditDetailsDialogVisible = wrapper.instance().state
      .isEditDetailsDialogVisible;
    wrapper.find('.edit-details-btn').simulate('click');
    expect(wrapper.instance().state.isEditDetailsDialogVisible).not.toBe(
      prevIsEditDetailsDialogVisible
    );
  });
});
