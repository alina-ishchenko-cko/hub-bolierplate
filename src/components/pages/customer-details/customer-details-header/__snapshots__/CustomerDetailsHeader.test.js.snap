// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customer Details Header renders correctly 1`] = `
.css-0 {
  width: 100%;
  padding-bottom: 30px;
}

.css-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  line-height: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: 0;
  padding: 0;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: auto;
}

.css-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  line-height: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: 0;
  padding: 0;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: auto;
  padding: 1.875rem;
}

.css-2 .customer-name {
  line-height: 1;
  color: #366BCF;
  font-size: 3.5384615384615383em;
}

.css-2 .actions {
  text-align: right;
}

.css-3 {
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: 0;
  padding: 0;
}

.css-4 {
  margin: 0;
  font-weight: 400;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-appearance: button;
  -moz-appearance: button;
  appearance: button;
  border-radius: 3px;
  text-align: center;
  border: solid 1px transparent;
  padding: 0 1.25rem;
  font-size: 1em;
  line-height: 42px;
  width: auto;
  cursor: pointer;
  color: #366bcf;
  background: #fff;
  box-shadow: 0px 1px 3px 0 rgba(121,131,149,0.15);
  border-color: #dfe1ea;
}

.css-4 .cko-icon svg * {
  fill: #366bcf;
}

.css-4:hover {
  box-shadow: 0px 2px 5px 0 rgba(121,131,149,0.25);
  border-color: #4780ec;
}

.css-4:active {
  background-color: #f5f8ff;
  box-shadow: inset 0px 2px 5px 0 rgba(121,131,149,0.2);
  border-color: #4780ec;
}

.css-4:focus {
  outline: none;
}

.css-4 .cko-icon {
  margin-right: 15px;
}

.css-8 {
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  text-transform: none;
  line-height: 1px;
  width: auto;
  border-radius: 3px;
}

.css-8.placeholder {
  background-color: #dfe1ea;
}

.css-8 svg {
  width: auto !important;
  height: auto !important;
  max-width: 100%;
}

.css-8 .white {
  fill: #fff;
}

.css-8 .dark {
  fill: #8b94a9;
}

.css-8 .light {
  fill: #bac0ce;
}

.css-8 .dark,
.css-8 .light,
.css-8 .dark,
.css-8 .grey {
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
}

.css-8 .icon-sidebar-transactions .dark,
.css-8 .icon-sidebar-dashboard .dark {
  fill: #8b94a9;
}

.css-8 .icon-sidebar-transactions .light,
.css-8 .icon-sidebar-dashboard .light {
  fill: #bac0ce;
}

.css-8 .icon-sidebar-plans .grey {
  fill: #fff;
}

.css-8 .icon-calendar * {
  fill: #ACB2BF;
}

.css-8 .icon-right-arrow *,
.css-8 .icon-left-arrow *,
.css-8 .icon-down-arrow * {
  fill: #abb3c4;
}

.css-8 .icon-paypal-full * {
  fill: #2464b4;
}

.css-10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  line-height: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: 0;
  padding: 0;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: auto;
  color: #798395;
  padding-left: 30px;
}

.css-10 .customer-info-column {
  margin-right: 50px;
}

.css-10 .email-icon {
  margin-right: 3px;
}

.css-10 .date-icon {
  margin-right: 6px;
}

.css-10 .phone-icon {
  margin-right: 6px;
}

.css-10 .id-icon {
  margin-right: 6px;
}

.css-10 .id-text {
  margin-right: 10px;
}

.css-11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  line-height: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  margin: 0;
  padding: 0;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: auto;
}

.css-12 {
  margin: 0;
  font-weight: 400;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-appearance: button;
  -moz-appearance: button;
  appearance: button;
  border-radius: 3px;
  text-align: center;
  border: solid 1px transparent;
  padding: 0 0.625rem;
  font-size: 0.9230769230769231em;
  line-height: 26px;
  width: auto;
  cursor: pointer;
  color: #366bcf;
  background: #fff;
  box-shadow: 0px 1px 3px 0 rgba(121,131,149,0.15);
  border-color: #dfe1ea;
}

.css-12 .cko-icon svg * {
  fill: #366bcf;
}

.css-12:hover {
  box-shadow: 0px 2px 5px 0 rgba(121,131,149,0.25);
  border-color: #4780ec;
}

.css-12:active {
  background-color: #f5f8ff;
  box-shadow: inset 0px 2px 5px 0 rgba(121,131,149,0.2);
  border-color: #4780ec;
}

.css-12:focus {
  outline: none;
}

.css-12 .cko-icon {
  margin-right: 10px;
}

<div
  className="customer-details-header css-0"
>
  <div
    className="cko-flex-column  css-1"
  >
    <div
      className="cko-flex-row css-2"
    >
      <div
        className="cko-flex-item customer-name css-3"
      >
        Alice
      </div>
      <div
        className="cko-flex-item actions css-3"
      >
        <button
          className="cko-button edit-details-btn css-4"
          onClick={[Function]}
          size="large"
          value="Edit"
        >
          <span
            className="cko-icon  css-8"
          >
            <span
              src="/assets/svgs/edit.svg"
            />
          </span>
          <span
            className="value"
          >
            Edit
          </span>
        </button>
      </div>
    </div>
    <div
      className="cko-flex-row css-10"
    >
      <div
        className="cko-flex-column customer-info-column css-11"
      >
        <div
          className="cko-flex-item email css-3"
        >
          <span
            className="cko-icon email-icon css-8"
          >
            <span
              src="/assets/svgs/email.svg"
            />
          </span>
          <span>
             
            a@a.com
          </span>
        </div>
        <div
          className="cko-flex-item create-date css-3"
        >
          <span
            className="cko-icon date-icon css-8"
          >
            <span
              src="/assets/svgs/date.svg"
            />
          </span>
          <span>
            17/04/2018 13:51:49
          </span>
        </div>
      </div>
      <div
        className="cko-flex-column customer-info-column css-11"
      >
        <div
          className="cko-flex-item phone css-3"
        >
          <span
            className="cko-icon phone-icon css-8"
          >
            <span
              src="/assets/svgs/phone.svg"
            />
          </span>
          <span>
            N/A
          </span>
        </div>
        <div
          className="cko-flex-item plan-id css-3"
        >
          <span
            className="cko-icon id-icon css-8"
          >
            <span
              src="/assets/svgs/id.svg"
            />
          </span>
          <span
            className="id-text"
          >
            plan_1
          </span>
          <button
            className="cko-button  css-12"
            onClick={[Function]}
            size="small"
            value="Copy"
          >
            <span
              className="value"
            >
              Copy
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
